File: src/main/java/lekkit/scev/client/DisplayState.java
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package lekkit.scev.client;

import com.mojang.blaze3d.platform.NativeImage;
import java.nio.ByteBuffer;
import java.util.UUID;
import lekkit.scev.machine.FramebufferView;
import lekkit.scev.main.ScalarEvolution;
import lekkit.scev.server.MachineState;
import net.minecraft.client.Minecraft;
import net.minecraft.client.renderer.texture.DynamicTexture;
import net.minecraft.resources.ResourceLocation;

/**
 * Per-machine framebuffer that's uploaded to a {@link DynamicTexture}. Supports two modes:
 * <ul>
 *   <li><b>Single-player (local)</b>: reads straight from the server-side
 *       {@link MachineState#getDisplay()}'s DMA buffer every frame.</li>
 *   <li><b>Multiplayer (remote)</b>: accepts pixel blobs through
 *       {@link #updateRemoteBuffer(byte[])}.</li>
 * </ul>
 *
 * <p>Pixel format: the source buffer is little-endian A8R8G8B8 (byte order
 * {@code B, G, R, A}); the destination {@link NativeImage} stores bytes as
 * {@code R, G, B, A}. {@link PixelConverter#bgraToRgba} does the swap.
 */
public class DisplayState {
    private final UUID uuid;
    private final int width;
    private final int height;

    private final byte[] remoteBuffer;           // only used when !isLocal()
    private final MachineState localMachine;     // only used when isLocal()

    private NativeImage image;
    private DynamicTexture texture;
    private ResourceLocation textureId;
    private boolean dirty = true;

    DisplayState(UUID uuid, int width, int height) {
        this.uuid = uuid;
        this.width = width;
        this.height = height;
        this.remoteBuffer = new byte[width * height * 4];
        this.localMachine = null;
    }

    DisplayState(MachineState machine) {
        FramebufferView fb = machine.getDisplay();
        if (fb == null) throw new IllegalArgumentException("MachineState has no display");
        this.uuid = machine.getUUID();
        this.width = fb.width();
        this.height = fb.height();
        this.remoteBuffer = null;
        this.localMachine = machine;
    }

    public UUID getUuid() { return uuid; }
    public int getWidth() { return width; }
    public int getHeight() { return height; }
    public boolean isLocal() { return localMachine != null; }

    synchronized void updateRemoteBuffer(byte[] src) {
        if (remoteBuffer == null || src.length != remoteBuffer.length) return;
        System.arraycopy(src, 0, remoteBuffer, 0, src.length);
        dirty = true;
    }

    /**
     * Returns the texture id for this framebuffer, uploading pixels if dirty.
     * Must be called on the render thread.
     */
    public synchronized ResourceLocation getOrUploadTexture() {
        if (texture == null) {
            image = new NativeImage(NativeImage.Format.RGBA, width, height, false);
            texture = new DynamicTexture(image);
            textureId = ScalarEvolution.rl("display/" + uuid.toString().replace('-', '_'));
            Minecraft.getInstance().getTextureManager().register(textureId, texture);
        }
        if (dirty || isLocal()) {
            refreshPixels();
            texture.upload();
            dirty = false;
        }
        return textureId;
    }

    private void refreshPixels() {
        ByteBuffer src;
        if (isLocal()) {
            FramebufferView fb = localMachine.getDisplay();
            if (fb == null) return;
            src = fb.pixels();
        } else {
            src = ByteBuffer.wrap(remoteBuffer);
        }
        PixelConverter.bgraToRgba(src, image, width, height);
    }

    public synchronized void destroy() {
        if (texture != null) texture.close();
        image = null;
        texture = null;
    }
}

File: src/main/kotlin/lekkit/scev/client/DisplayState.kt
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package lekkit.scev.client

import com.mojang.blaze3d.platform.NativeImage
import java.nio.ByteBuffer
import java.util.UUID
import lekkit.scev.machine.FramebufferView
import lekkit.scev.main.ScalarEvolution
import lekkit.scev.server.MachineState
import net.minecraft.client.Minecraft
import net.minecraft.client.renderer.texture.DynamicTexture
import net.minecraft.resources.ResourceLocation

/**
 * Per-machine framebuffer uploaded to a [DynamicTexture]. Two modes:
 *
 *  - **Single-player (local):** reads straight from the server-side
 *    [MachineState.getDisplay] DMA buffer every frame.
 *  - **Multiplayer (remote):** accepts pixel blobs through
 *    [updateRemoteBuffer].
 *
 * Pixel format: source is little-endian A8R8G8B8 (byte order `B, G, R, A`);
 * destination [NativeImage] stores bytes as `R, G, B, A`. [PixelConverter.bgraToRgba]
 * does the swap.
 */
class DisplayState private constructor(
    @get:JvmName("getUuid")   val uuid: UUID,
    @get:JvmName("getWidth")  val width: Int,
    @get:JvmName("getHeight") val height: Int,
    private val remoteBuffer: ByteArray?,
    private val localMachine: MachineState?,
) {
    private var image: NativeImage? = null
    private var texture: DynamicTexture? = null
    private var textureId: ResourceLocation? = null
    private var dirty: Boolean = true

    /** Constructor for the multiplayer / remote case. */
    internal constructor(uuid: UUID, width: Int, height: Int) :
        this(uuid, width, height, ByteArray(width * height * 4), null)

    /** Constructor for the singleplayer / local-machine case. */
    internal constructor(machine: MachineState) : this(
        machine.getUuid(),
        machine.display?.width() ?: throw IllegalArgumentException("MachineState has no display"),
        machine.display!!.height(),
        null,
        machine,
    )

    fun isLocal(): Boolean = localMachine != null

    /**
     * True if this is a singleplayer DisplayState whose backing
     * [MachineState] has been destroyed (VM was powered off). The cache
     * entry is dead; [DisplayManager.get] evicts stale entries before
     * returning so the next lookup can construct fresh against the new
     * MachineState (or return null if no VM is running for the UUID).
     */
    val isStale: Boolean
        get() = isLocal() && !localMachine!!.isValid

    @Synchronized internal fun updateRemoteBuffer(src: ByteArray) {
        val buf = remoteBuffer ?: return
        if (src.size != buf.size) return
        System.arraycopy(src, 0, buf, 0, src.size)
        dirty = true
    }

    /**
     * Returns the texture id for this framebuffer, uploading pixels if
     * dirty. Must be called on the render thread.
     */
    @Synchronized fun getOrUploadTexture(): ResourceLocation {
        if (texture == null) {
            val img = NativeImage(NativeImage.Format.RGBA, width, height, false)
            image = img
            texture = DynamicTexture(img)
            textureId = ScalarEvolution.rl("display/" + uuid.toString().replace('-', '_'))
            Minecraft.getInstance().textureManager.register(textureId, texture)
        }
        if (dirty || isLocal()) {
            refreshPixels()
            texture!!.upload()
            dirty = false
        }
        return textureId!!
    }

    private fun refreshPixels() {
        val src: ByteBuffer = if (isLocal()) {
            val fb: FramebufferView? = localMachine!!.display
            if (fb == null) {
                // VM was torn down (power-off, chunk-unload free-all, etc.).
                // Paint the texture black so the last frame from the old VM
                // doesn't linger — stale DisplayStates are evicted from
                // DisplayManager on the next get(), but a mid-frame clear
                // avoids the "screen shows prior boot" flash when the user
                // re-opens the MachineScreen.
                clearPixels()
                return
            }
            fb.pixels()
        } else {
            ByteBuffer.wrap(remoteBuffer!!)
        }
        PixelConverter.bgraToRgba(src, image!!, width, height)
    }

    /** Zero every pixel in the NativeImage (opaque black). */
    private fun clearPixels() {
        val img = image ?: return
        for (y in 0 until height) for (x in 0 until width) img.setPixelRGBA(x, y, 0xFF000000.toInt())
    }

    @Synchronized fun destroy() {
        texture?.close()
        image = null
        texture = null
    }
}
